lib/pbench/test/unit/server/test_datasets_list.py

```diff
@@ -356,3 +380,69 @@ def test_filter_query(self, monkeypatch, filters, expected):
             )
             == prefix + expected
         )
+
+    def test_user_no_auth(self, monkeypatch, db_session):
+        """Test the authorization error when a match against a key in the user
+        namespace is attempted from an unauthenticated session.
+        """
+        monkeypatch.setattr(
+            "pbench.server.api.resources.datasets_list.Auth.get_current_user_id",
+            lambda: None,
+        )
+        with pytest.raises(APIAbort) as e:
+            DatasetsList.filter_query(
+                ["user.foo:1"], Database.db_session.query(Dataset).outerjoin(Metadata)
+            )
+        assert e.value.http_status == HTTPStatus.UNAUTHORIZED
+
+    @pytest.mark.parametrize(
+        "meta,error",
+        [
+            ("user.foo:1", HTTPStatus.UNAUTHORIZED),
+            ("x.y:1", HTTPStatus.BAD_REQUEST),
+            ("global.x=3", HTTPStatus.BAD_REQUEST),
+            ("dataset.notright:10", HTTPStatus.BAD_REQUEST),
+        ],
+    )
+    def test_filter_errors(self, monkeypatch, db_session, meta, error):
+        """Test invalid filter expressions."""
+        monkeypatch.setattr(
+            "pbench.server.api.resources.datasets_list.Auth.get_current_user_id",
+            lambda: None,
+        )
+        with pytest.raises(APIAbort) as e:
+            DatasetsList.filter_query(
+                [meta], Database.db_session.query(Dataset).outerjoin(Metadata)
+            )
+        assert e.value.http_status == error
+
+    @pytest.mark.parametrize(
+        "exception,error",
+        (
+            (Exception("test"), "Unexpected SQL exception: test"),
+            (
+                ProgrammingError("stmt", [], "orig"),
+                "Constructed SQL for {} isn't executable",
+            ),
+        ),
+    )
+    def test_pagination_error(self, caplog, monkeypatch, query_as, exception, error):
+        """Test that query problems during pagination are reported as server
+        internal errors.
+        """
+
+        def do_error(
+            self, query: Query, json: JSONOBJECT, url: str
+        ) -> tuple[JSONARRAY, JSONOBJECT]:
+            raise exception
+
+        monkeypatch.setattr(DatasetsList, "get_paginated_obj", do_error)
+        response = query_as({}, "drb", HTTPStatus.INTERNAL_SERVER_ERROR)
+        message = response.json["message"]
+        match = LOG_SEQ.match(message)
+        key = match.group(1)
+        assert match and key, "Error response {message!r} is not an internal error"
```

Comment:
Not to be picky or anything, but shouldn't we be checking `match` before we invoke `group()` on it?  (Otherwise, we risk incurring an `AttributeError` before we get to the nice `assert` message if the match comes up empty.)

Reply:
I considered that and decided not to bother. It's a less clear error, sure, but it's unlikely and I didn't see any point in worrying about it.